File: examples/python/m2-tool-runtime/tools/grep.py
"""grep：在工作区内搜索文件内容。"""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any, Mapping

from tool_types import ToolSpec
from tools.glob_match import glob_matches
from workspace import Workspace, WorkspaceError

DEFAULT_LIMIT = 100
MAX_FILE_BYTES = 1_000_000


class GrepTool:
    """简单 Python 实现的内容搜索；不依赖外部 rg。"""

    def __init__(self, workspace: Workspace) -> None:
        self.workspace = workspace
        self.spec = ToolSpec(
            name="grep",
            description="Search file contents inside the workspace",
            input_schema={
                "type": "object",
                "required": ["pattern"],
                "properties": {
                    "pattern": {"type": "string"},
                    "path": {"type": "string"},
                    "glob": {"type": "string"},
                    "ignoreCase": {"type": "boolean"},
                    "literal": {"type": "boolean"},
                    "context": {"type": "integer"},
                    "limit": {"type": "integer"},
                },
                "additionalProperties": False,
            },
        )

    def validate_arguments(self, arguments: Mapping[str, Any]) -> None:
        pattern = arguments.get("pattern")
        if not isinstance(pattern, str) or pattern == "":
            raise ValueError("pattern must be a non-empty string")
        for key in ("path", "glob"):
            value = arguments.get(key)
            if value is not None and not isinstance(value, str):
                raise ValueError(f"{key} must be a string")
        for key in ("ignoreCase", "literal"):
            value = arguments.get(key)
            if value is not None and not isinstance(value, bool):
                raise ValueError(f"{key} must be a boolean")
        for key in ("context", "limit"):
            value = arguments.get(key)
            if value is not None and (
                not isinstance(value, int) or isinstance(value, bool) or value < 0
            ):
                raise ValueError(f"{key} must be a non-negative integer")

    def execute(self, arguments: Mapping[str, Any]) -> dict[str, Any]:
        root = self.workspace.resolve(arguments.get("path"), must_exist=True)
        pattern_text = str(arguments["pattern"])
        literal = bool(arguments.get("literal", False))
        ignore_case = bool(arguments.get("ignoreCase", False))
        context = int(arguments.get("context") or 0)
        limit = int(arguments.get("limit") or DEFAULT_LIMIT)
        glob_pat = arguments.get("glob") or "**/*"

        flags = re.MULTILINE
        if ignore_case:
            flags |= re.IGNORECASE
        try:
            regex = (
                re.compile(re.escape(pattern_text), flags)
                if literal
                else re.compile(pattern_text, flags)
            )
        except re.error as error:
            raise ValueError(f"invalid regex: {error}") from error

        files = self._iter_files(root, str(glob_pat))
        matches: list[dict[str, Any]] = []
        truncated = False

        for file_path in files:
            if len(matches) >= limit:
                truncated = True
                break
            try:
                if file_path.stat().st_size > MAX_FILE_BYTES:
                    continue
                text = file_path.read_text(encoding="utf-8")
            except (OSError, UnicodeDecodeError):
                continue

            lines = text.splitlines()
            rel = self.workspace.relative_to_root(file_path)
            for line_no, line in enumerate(lines, start=1):
                if not regex.search(line):
                    continue
                start = max(0, line_no - 1 - context)
                end = min(len(lines), line_no + context)
                snippet = "\n".join(
                    f"{idx + 1}: {lines[idx]}" for idx in range(start, end)
                )
                matches.append(
                    {
                        "path": rel,
                        "line": line_no,
                        "text": line,
                        "snippet": snippet,
                    }
                )
                if len(matches) >= limit:
                    truncated = True
                    break

        return {
            "pattern": pattern_text,
            "matches": matches,
            "count": len(matches),
            "truncated": truncated,
        }

    def _iter_files(self, root: Path, glob_pat: str) -> list[Path]:
        if root.is_file():
            return [root]
        if not root.is_dir():
            raise WorkspaceError(f"not a file or directory: {root}")
        results: list[Path] = []
        # 与 Rust 一样先递归枚举，再使用共享的 glob 子集匹配相对路径。
        for path in sorted(root.rglob("*")):
            if path.is_file():
                self.workspace.resolve(str(path))
                relative = path.resolve().relative_to(root).as_posix()
                if glob_matches(glob_pat, relative):
                    results.append(path)
        return results

File: examples/rust/m2-tool-runtime/src/tools/grep.rs
use crate::registry::{require_object, require_string};
use crate::tool_types::{Tool, ToolError, ToolSpec};
use crate::workspace::Workspace;
use regex::RegexBuilder;
use serde_json::{Value, json};
use std::path::PathBuf;

const DEFAULT_LIMIT: usize = 100;
const MAX_FILE_BYTES: u64 = 1_000_000;

/// 在 Workspace 文本文件中执行正则或字面量搜索。
pub struct GrepTool {
    workspace: Workspace,
}

impl GrepTool {
    /// 绑定固定 Workspace。
    pub fn new(workspace: Workspace) -> Self {
        Self { workspace }
    }
}

impl Tool for GrepTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "grep".into(),
            description: "Search file contents inside the workspace".into(),
            input_schema: json!({
                "type": "object",
                "required": ["pattern"],
                "properties": {
                    "pattern": {"type": "string"},
                    "path": {"type": "string"},
                    "glob": {"type": "string"},
                    "ignoreCase": {"type": "boolean"},
                    "literal": {"type": "boolean"},
                    "context": {"type": "integer"},
                    "limit": {"type": "integer"}
                },
                "additionalProperties": false
            }),
            strict: false,
        }
    }

    fn validate_arguments(&self, arguments: &Value) -> Result<(), ToolError> {
        let object = require_object(arguments)?;
        let pattern = object
            .get("pattern")
            .and_then(Value::as_str)
            .ok_or_else(|| ToolError::new("pattern must be a non-empty string"))?;
        if pattern.is_empty() {
            return Err(ToolError::new("pattern must be a non-empty string"));
        }
        for key in ["path", "glob"] {
            if object.get(key).is_some_and(|value| !value.is_string()) {
                return Err(ToolError::new(format!("{key} must be a string")));
            }
        }
        if object
            .get("glob")
            .and_then(Value::as_str)
            .is_some_and(|glob| glob.contains('[') || glob.contains(']'))
        {
            return Err(ToolError::new("glob character classes are not supported"));
        }
        for key in ["ignoreCase", "literal"] {
            if object.get(key).is_some_and(|value| !value.is_boolean()) {
                return Err(ToolError::new(format!("{key} must be a boolean")));
            }
        }
        for key in ["context", "limit"] {
            if object
                .get(key)
                .is_some_and(|value| value.as_u64().is_none())
            {
                return Err(ToolError::new(format!(
                    "{key} must be a non-negative integer"
                )));
            }
        }
        Ok(())
    }

    fn execute(&mut self, arguments: &Value) -> Result<Value, ToolError> {
        let object = require_object(arguments)?;
        let pattern = require_string(object, "pattern")?;
        let path_arg = object.get("path").and_then(Value::as_str);
        let root = self.workspace.resolve(path_arg, true)?;
        let ignore_case = object
            .get("ignoreCase")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        let literal = object
            .get("literal")
            .and_then(Value::as_bool)
            .unwrap_or(false);
        let context = object.get("context").and_then(Value::as_u64).unwrap_or(0) as usize;
        let limit = object
            .get("limit")
            .and_then(Value::as_u64)
            .map(|value| value as usize)
            .unwrap_or(DEFAULT_LIMIT);
        let glob = object.get("glob").and_then(Value::as_str).unwrap_or("**/*");

        let expression = if literal {
            regex::escape(pattern)
        } else {
            pattern.to_owned()
        };
        let regex = RegexBuilder::new(&expression)
            .case_insensitive(ignore_case)
            .multi_line(true)
            .build()
            .map_err(|error| ToolError::new(format!("invalid regex: {error}")))?;

        let files = collect_files(&root, glob)?;
        let mut matches = Vec::new();
        let mut truncated = false;
        for file in files {
            if matches.len() >= limit {
                truncated = true;
                break;
            }
            let meta = match std::fs::metadata(&file) {
                Ok(meta) => meta,
                Err(_) => continue,
            };
            if meta.len() > MAX_FILE_BYTES {
                continue;
            }
            let text = match std::fs::read_to_string(&file) {
                Ok(text) => text,
                Err(_) => continue,
            };
            let lines: Vec<&str> = text.lines().collect();
            let rel = self.workspace.relative_to_root(&file)?;
            for (index, line) in lines.iter().enumerate() {
                if !regex.is_match(line) {
                    continue;
                }
                let line_no = index + 1;
                let start = index.saturating_sub(context);
                let end = (index + 1 + context).min(lines.len());
                let snippet = lines[start..end]
                    .iter()
                    .enumerate()
                    .map(|(offset, item)| format!("{}: {item}", start + offset + 1))
                    .collect::<Vec<_>>()
                    .join("\n");
                matches.push(json!({
                    "path": rel,
                    "line": line_no,
                    "text": line,
                    "snippet": snippet,
                }));
                if matches.len() >= limit {
                    truncated = true;
                    break;
                }
            }
        }

        Ok(json!({
            "pattern": pattern,
            "matches": matches,
            "count": matches.len(),
            "truncated": truncated,
        }))
    }
}

fn collect_files(root: &std::path::Path, glob_pat: &str) -> Result<Vec<PathBuf>, ToolError> {
    if root.is_file() {
        return Ok(vec![root.to_path_buf()]);
    }
    if !root.is_dir() {
        return Err(ToolError::new(format!("not a file or directory: {root:?}")));
    }
    let mut files = Vec::new();
    let mut stack = vec![root.to_path_buf()];
    while let Some(dir) = stack.pop() {
        for entry in std::fs::read_dir(&dir).map_err(|error| ToolError::new(error.to_string()))? {
            let entry = entry.map_err(|error| ToolError::new(error.to_string()))?;
            let path = entry.path();
            if path.is_dir() {
                stack.push(path);
                continue;
            }
            if path.is_file() {
                let rel = path
                    .strip_prefix(root)
                    .unwrap_or(&path)
                    .to_string_lossy()
                    .replace('\\', "/");
                if crate::tools::find::glob_match(glob_pat, &rel) {
                    files.push(path);
                }
            }
        }
    }
    files.sort();
    Ok(files)
}
